keps/sig-apps/0034-maxunavailable-for-statefulsets.md

```diff
@@ -0,0 +1,179 @@
+
+### Implementation Details
+
+#### API Changes
+
+Following changes will be made to the Rolling Update Strategy for StatefulSet.
+
+```go
+// RollingUpdateStatefulSetStrategy is used to communicate parameter for RollingUpdateStatefulSetStrategyType.
+type RollingUpdateStatefulSetStrategy struct {
+        // Partition indicates the ordinal at which the StatefulSet should be
+        // partitioned.
+        // Default value is 0.
+        // +optional
+        Partition *int32 `json:"partition,omitempty" protobuf:"varint,1,opt,name=partition"`
+
+	// The maximum number of pods that can be unavailable during the update.
+        // Value can be an absolute number (ex: 5) or a percentage of desired pods (ex: 10%).
+        // Absolute number is calculated from percentage by rounding down.
+        // Defaults to 1.
+        // +optional
+        MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty" protobuf:"bytes,2,opt,name=maxUnavailable"`
+	
+	...
+}
+```
+
+- By Default, if maxUnavailable is not specified, its value will be assumed to be 1 and StatefulSets will follow their old behavior.
+- if MaxUnavailable is specified, it cannot be greater than total number of replicas
+- If a partition is specified, maxUnavailable will only apply to all the pods which are staged by the partition. Lets say total replicas
+is 5 and partition is set to 2 and maxUnavailable is set to 2. If the image is changed in this scenario, pods with ordinal 4 and 3 will go
+down at the same time(because of maxUnavailable), once they are running and ready, pods with ordinal 2 will go down. Pods with ordinal 0
+and 1 will remain untouched due the partition.
+
+#### Implementation
+
+https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_control.go#L504
+```go
+...
+	podsDeleted := 0
+	// we terminate the Pod with the largest ordinal that does not match the update revision.
+	for target := len(replicas) - 1; target >= updateMin; target-- {
+
+		// delete the Pod if it is not already terminating and does not match the update revision.
+		if getPodRevision(replicas[target]) != updateRevision.Name && !isTerminating(replicas[target]) {
+			klog.V(2).Infof("StatefulSet %s/%s terminating Pod %s for update",
+				set.Namespace,
+				set.Name,
+				replicas[target].Name)
+			err := ssc.podControl.DeleteStatefulPod(set, replicas[target])
+			status.CurrentReplicas--
+
+			if podsDeleted < set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable {
+				podsDeleted ++;
+				continue;
+			}
+			return &status, err
+		}
+
+		// wait for unhealthy Pods on update
+		if !isHealthy(replicas[target]) {
+			klog.V(4).Infof(
+				"StatefulSet %s/%s is waiting for Pod %s to update",
+				set.Namespace,
+				set.Name,
+				replicas[target].Name)
+			return &status, nil
+		}
+
+	}
+...
+```
+
+### Risks and Mitigations
+
+## Graduation Criteria
```

Comment:
Please add a plan for implementation and delivery. Use your best estimate for when you think you can implement each phase of the modification. Consider how/if the feature will be gated at alpha. What will the criteria for beta and GA be? Once enabled, can it be rolled back safely?